### PR TITLE
obsservice: ability to produce events into event queues

### PR DESCRIPTION
**Pull request by abarganier**
Once an event in the obsservice has been routed, it is ready
to be transformed into its event-specific type, validated, and
then produced into an event queue to await aggregation.

While the individual implementations for how to transform,
validate, and enqueue events varies across event types and
event queue implementations, the common group of operations
can be generalized into a common ProducerGroup implementation,
with the transform, validate, and produce logic injected at
initialization.

This PR defines the necessary interfaces for this, as well
as the common ProducerGroup logic.

Additionally, we define an in-memory queue implementation 
(`MemoryQueue`) and associated EventProducer implementation 
to use as our first means of buffering events for aggregation. 

In a future batch of work, where we scaffold these interfaces for
event aggregation, an associated consumer for the `MemoryQueue`
will be defined to feed queued events to aggregators.

Release note: none

Epic: CRDB-28191

Addresses: https://github.com/cockroachdb/cockroach/issues/104811

---
**Reviewer note:** recommended to review commit-wise.

### CODE REVIEW

**Comment by abarganier** [comment]
@j82w I'm adding you as a reviewer for visibility & context, since the transformer & validator interfaces will need to be implemented for `eventpb.SampledQuery` events when implementing the SQL stats feature in the obsservice.

**Review by knz — APPROVED** [review]
> Reviewed 9 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107191)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, and @j82w)

___
*[`pkg/obsservice/obslib/produce/producer.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-N_u929h1De58J-DheFD:-N_u929h1De58J-DheFE:bxdc5jg) ([raw file](https://github.com/cockroachdb/cockroach/blob/67270a5fe66f6583422a94ac389e6943b0e0eb61/pkg/obsservice/obslib/produce/producer.go#L53)):*
> ```Go
> 		return nil, errors.New("nil transformer provided")
> 	}
> 	if validators == nil {
> ```

don't you want to check `len(validators) == 0` instead? An empty slice will not compare equal to `nil`.
ditto producers

**Review by abarganier — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107191)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @j82w, and @knz)

___
*[`pkg/obsservice/obslib/produce/producer.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-N_u929h1De58J-DheFD:-Na7KVyV8ZyMMV6fKeOv:b5g78tj) ([raw file](https://github.com/cockroachdb/cockroach/blob/67270a5fe66f6583422a94ac389e6943b0e0eb61/pkg/obsservice/obslib/produce/producer.go#L53)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

don't you want to check `len(validators) == 0` instead? An empty slice will not compare equal to `nil`.
ditto producers
</blockquote></details>

Oops, good catch, thanks - switched to a length check.

File: pkg/obsservice/obslib/queue/mem_queue.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+// to enforce the configured maxSize.
+//
+// MemoryQueue is backed by the container/list package from the Golang
+// stdlib, which is a linked list under the hood. This allows us to avoid
+// holding onto unused memory like you would by trimming a pre-allocated
+// slice (e.g. mySlice[1:]) when dequeue'ing elements from the front.
+// See: https://pkg.go.dev/container/list for details.
+//
+// See NewMemoryQueue for initialization.
+type MemoryQueue[T proto.Sizer] struct {
```
Comment (koorosh) [inline pkg/obsservice/obslib/queue/mem_queue.go:33]: `MemoryQueue` implementation seems to be very universal and not limited to `obsservice` only. Would it be appropriate to move it to `pkg/util`? (it already has  `pkg/util/cache` package that provides in memory cache with with specific restrictions so this one could be a one more)

**Review by j82w — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107191)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, and @knz)

___
*[`-- commits` line 8 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NaSRejx2xVdaf60jKC3):*
Should it be validate and then transform? That way it avoid the overhead of transforming if the object is not valid. It also avoids adding a bunch of validation code to the transformer.
___
*[`pkg/obsservice/obslib/produce/mem_queue_producer.go` line 39 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NaSHu0N-ve5yzhIe1xH:-NaSHu0N-ve5yzhIe1xI:bqfxx4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0b87e29a8a76b2b202e5778b37aae6cac03427/pkg/obsservice/obslib/produce/mem_queue_producer.go#L39)):*
> ```Go
> }
> 
> func (m *MemQueueProducer[T]) Produce(e interface{}) error {
> ```

I'm not that familiar with go generics, but why not use T instead of interface to avoid the cast? `Produce(e T) error {`
___
*[`pkg/obsservice/obslib/produce/producer.go` line 39 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NaSLA-jBtL8Z1oC_yU6:-NaSLA-jBtL8Z1oC_yU7:b-btnfaj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d831aed2f47b4e5308e1341be8d35835cecc595/pkg/obsservice/obslib/produce/producer.go#L39)):*
> ```Go
> 	transformer transform.EventTransformer
> 	validators  []validate.Validator
> 	producers   []EventProducer
> ```

Is there any way to make the transformer, validator, and producer use the same type via generics? It seems painful and error prone to do type checking in all the implementations.
___
*[`pkg/obsservice/obslib/validate/validate.go` line 13 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NaSJxrn-F1x7HBoH2OC:-NaSJxrn-F1x7HBoH2OD:b47bm09) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d831aed2f47b4e5308e1341be8d35835cecc595/pkg/obsservice/obslib/validate/validate.go#L13)):*
> ```Go
> // Validator is the interface used to validate
> // event-specific types.
> type Validator interface {
> ```

It's unclear to me is when and which interfaces needs to be thread safe vs not. Should this be called out or labeled somewhere to avoid bugs?

**Review by abarganier — COMMENTED** [review]
Long time no see! This PR is being revived & driven home.

@j82w PTAL when you have a moment.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-:-NfvlTl741pH5kW9X6Lw:bvttk0c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @j82w, @knz, and @koorosh)

___
*[`pkg/obsservice/obslib/produce/mem_queue_producer.go` line 39 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NaSHu0N-ve5yzhIe1xH:-NfvPmkS6roJ0WL_smyA:b6oa6co) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0b87e29a8a76b2b202e5778b37aae6cac03427/pkg/obsservice/obslib/produce/mem_queue_producer.go#L39)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

I'm not that familiar with go generics, but why not use T instead of interface to avoid the cast? `Produce(e T) error {`
</blockquote></details>

After picking this PR back up, I agree. Switched to using the same generic as the mem queue.
___
*[`pkg/obsservice/obslib/produce/producer.go` line 39 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NaSLA-jBtL8Z1oC_yU6:-NfvPtBGDgpP4s1EVapZ:bnvxr3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d831aed2f47b4e5308e1341be8d35835cecc595/pkg/obsservice/obslib/produce/producer.go#L39)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

Is there any way to make the transformer, validator, and producer use the same type via generics? It seems painful and error prone to do type checking in all the implementations.
</blockquote></details>

Yeah, after revisiting this I completely agree. I've switched each to follow the same generic type, meaning no more type assertions. 
___
*[`pkg/obsservice/obslib/queue/mem_queue.go` line 33 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107191#gh-1277565474:-NfvQJWa0RrjaKONuuNC:bdpq9xf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0b87e29a8a76b2b202e5778b37aae6cac03427/pkg/obsservice/obslib/queue/mem_queue.go#L33)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

`MemoryQueue` implementation seems to be very universal and not limited to `obsservice` only. Would it be appropriate to move it to `pkg/util`? (it already has  `pkg/util/cache` package that provides in memory cache with with specific restrictions so this one could be a one more)
</blockquote></details>

I think this would be a good idea at some point, but for now I'd like to keep things obsservice specific in the interest of speed for prototyping (and avoiding other folks creating dependencies on a library that I'm not committed to keeping general in use/not specific to obsservice). 
___
*[`pkg/obsservice/obslib/validate/validate.go` line 13 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NaSJxrn-F1x7HBoH2OC:-Nfvj5Xw6hiJynbzksrS:b-p3wavx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d831aed2f47b4e5308e1341be8d35835cecc595/pkg/obsservice/obslib/validate/validate.go#L13)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

It's unclear to me is when and which interfaces needs to be thread safe vs not. Should this be called out or labeled somewhere to avoid bugs?
</blockquote></details>

Good point. I expect all to be thread safe (a single ProducerGroup instance can be called by multiple goroutines).

I've updated the docs on all interfaces to note this.

**Comment by j82w** [comment]
*[`pkg/obsservice/obslib/produce/producer.go` line 28 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NfvvvWCEQy1mlgoNsZZ:-NfvvvWCEQy1mlgoNsZ_:bps6gd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/001d017d93fde5c63945e82fd04c91543348ebdb/pkg/obsservice/obslib/produce/producer.go#L28)):*
> ```Go
> // EventProducer implementations are expected to be thread-safe.
> type EventProducer[T any] interface {
> 	Produce(T) error
> ```

Would it be useful if validators, producers, and transformers had a name or owner to see which one failed and who owns it? Not sure if there is a better way to get the info.

**Review by abarganier — COMMENTED** [review]
Thanks for the feedback. Fixed a small linter error.

@j82w PTAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-:-NfwIGbq6ETLRGCadmBl:b-3vno01)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @j82w, @knz, and @koorosh)

___
*[`-- commits` line 8 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NaSRejx2xVdaf60jKC3):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

Should it be validate and then transform? That way it avoid the overhead of transforming if the object is not valid. It also avoids adding a bunch of validation code to the transformer.
</blockquote></details>

Sorry I missed this. 

Sounds like a need for two types of validation. Prior to transformation, I think the only things we can validate are the metadata associated with the event (things like presence checks for node ID, event type, etc). Generally, this metadata should be fairly consistent across event types.

The EventRouter (a stage before the ProducerGroup) makes presence checks for event type, which is a requirement to route the event. Perhaps we should expand those checks to validate other metadata that's not event-specific. We could create a single, event-type-agnostic implementation that gets used in every `ProducerGroup` regardless of the type, and call it before we invoke the transformer.

LMK what you think.
___
*[`pkg/obsservice/obslib/produce/producer.go` line 28 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NfvvvWCEQy1mlgoNsZZ:-NfvxzF-3KrkAA83lXTL:b-5blpr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/001d017d93fde5c63945e82fd04c91543348ebdb/pkg/obsservice/obslib/produce/producer.go#L28)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

Would it be useful if validators, producers, and transformers had a name or owner to see which one failed and who owns it? Not sure if there is a better way to get the info.
</blockquote></details>

In my head, the validators/producers/transformers don't exist without belonging to a `ProducerGroup` instance, which has an alias of its own. If any validator, producer, or transformer being used in a ProducerGroup encounters an error, it's logged and the message contains the alias of the ProducerGroup. Do you think that's sufficient?

For example:
```
log.Errorf(ctx, "%s failed to transform event %v, err = %v",  p.alias, event, err)
```

**Comment by j82w** [comment]
*[`-- commits` line 8 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NaSRejx2xVdaf60jKC3):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

Sorry I missed this. 

Sounds like a need for two types of validation. Prior to transformation, I think the only things we can validate are the metadata associated with the event (things like presence checks for node ID, event type, etc). Generally, this metadata should be fairly consistent across event types.

The EventRouter (a stage before the ProducerGroup) makes presence checks for event type, which is a requirement to route the event. Perhaps we should expand those checks to validate other metadata that's not event-specific. We could create a single, event-type-agnostic implementation that gets used in every `ProducerGroup` regardless of the type, and call it before we invoke the transformer.

LMK what you think.
</blockquote></details>

I think a single agnostic implementation should work. I think adding to many options at the beginning adds unnecessary overhead and complexity. It should be easy enough to add additional validation in the future if necessary.

**Review by j82w — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107191)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @knz, and @koorosh)

___
*[`pkg/obsservice/obslib/produce/producer.go` line 39 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-Ng14zM20oGBWKLUnwHa:-Ng14zM20oGBWKLUnwHb:b8j7j4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/99fb458faae044baadb8e4d8f3e2ab4bee1ad370/pkg/obsservice/obslib/produce/producer.go#L39)):*
> ```Go
> //
> // ProducerGroup is thread-safe.
> type ProducerGroup[T any] struct {
> ```

Should the generic be based on some base struct that has event type and other basic properties that are required to process it?
___
*[`pkg/obsservice/obslib/produce/producer.go` line 88 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-Ng15He62TYyGZPnTmNK:-Ng15He62TYyGZPnTmNL:b-i6axub) ([raw file](https://github.com/cockroachdb/cockroach/blob/99fb458faae044baadb8e4d8f3e2ab4bee1ad370/pkg/obsservice/obslib/produce/producer.go#L88)):*
> ```Go
> 	}
> 	if validationErrors != nil {
> 		log.Errorf(ctx, "%s validation failed for event %v, err = %v",
> ```

I think this works for now, but might be problematic down the road. For example if a new cockroach version gets deployed and starts sending bad event it could cause a flood of error log messages which causes memory and other issues.
___
*[`pkg/obsservice/obslib/produce/producer.go` line 28 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NfvvvWCEQy1mlgoNsZZ:-Ng13_aq8N95tkb16M0g:b-hwpkug) ([raw file](https://github.com/cockroachdb/cockroach/blob/001d017d93fde5c63945e82fd04c91543348ebdb/pkg/obsservice/obslib/produce/producer.go#L28)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

In my head, the validators/producers/transformers don't exist without belonging to a `ProducerGroup` instance, which has an alias of its own. If any validator, producer, or transformer being used in a ProducerGroup encounters an error, it's logged and the message contains the alias of the ProducerGroup. Do you think that's sufficient?

For example:
```
log.Errorf(ctx, "%s failed to transform event %v, err = %v",  p.alias, event, err)
```
</blockquote></details>

I agree the `ProducerGroup` level is good.  Is `alias` a name for the producer group or a github `alias` to represent ownership? I think it might be nice for a ProducerGroup to have a owner that can be correlated to a team. That way down the road it's possible to figure out that backup has 5 `ProducerGroup`'s and is processing a million events a day across all the groups.

**Review by abarganier — COMMENTED** [review]
Thanks for the feedback @j82w - RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-:-Ng4aO9D4AWoE3zKsA2P:b-xbb3dq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @j82w, @knz, and @koorosh)

___
*[`-- commits` line 8 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NaSRejx2xVdaf60jKC3):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

I think a single agnostic implementation should work. I think adding to many options at the beginning adds unnecessary overhead and complexity. It should be easy enough to add additional validation in the future if necessary.
</blockquote></details>

Ack, sounds good to me. I'm going to mark this as a TODO for now. 
___
*[`pkg/obsservice/obslib/produce/producer.go` line 39 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-Ng14zM20oGBWKLUnwHa:-Ng4UctI6v3yqUWfkiFw:b-ipqvwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/99fb458faae044baadb8e4d8f3e2ab4bee1ad370/pkg/obsservice/obslib/produce/producer.go#L39)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

Should the generic be based on some base struct that has event type and other basic properties that are required to process it?
</blockquote></details>

`T` represents the type that the generic `obspb.Event` is transformed into. That type can be anything - the basic properties required to process it, at least for now, are those needed to route the event to the correct pipeline. For now, that lives in the `obspb.Event`'s metadata. 

I can see a need arising for what you're talking about, especially if things like friendly/cluster IDs get involved... But I don't have any concrete need at the moment. Marked as TODO for now! Baby steps :smile: 
___
*[`pkg/obsservice/obslib/produce/producer.go` line 88 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-Ng15He62TYyGZPnTmNK:-Ng4V_TU0RaR3ah_L7Qn:b2wzpl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/99fb458faae044baadb8e4d8f3e2ab4bee1ad370/pkg/obsservice/obslib/produce/producer.go#L88)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

I think this works for now, but might be problematic down the road. For example if a new cockroach version gets deployed and starts sending bad event it could cause a flood of error log messages which causes memory and other issues.
</blockquote></details>

Yeah agreed, I was thinking about this as well. We're going to need a smarter logging strategy here. Marked as a TODO for now. In the meantime, to cut down a bit on the volume, I've removed the logging of the event itself from each log call.
___
*[`pkg/obsservice/obslib/produce/producer.go` line 28 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NfvvvWCEQy1mlgoNsZZ:-Ng4Srl19Q1O6XXLNH-m:b8cpwy1) ([raw file](https://github.com/cockroachdb/cockroach/blob/001d017d93fde5c63945e82fd04c91543348ebdb/pkg/obsservice/obslib/produce/producer.go#L28)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

I agree the `ProducerGroup` level is good.  Is `alias` a name for the producer group or a github `alias` to represent ownership? I think it might be nice for a ProducerGroup to have a owner that can be correlated to a team. That way down the road it's possible to figure out that backup has 5 `ProducerGroup`'s and is processing a million events a day across all the groups.
</blockquote></details>

`alias` is a name for the producer group, which should be specific to an event type. 

I've added an additional field for an owner team, good idea. I also added a `logPrefix` field that's a combination of the two for us to use.

**Review by j82w — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107191)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @knz, and @koorosh)

___
*[`pkg/obsservice/obslib/produce/producer.go` line 68 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NgK2XxP0bgzGXzx-2rm:-NgK2XxP0bgzGXzx-2rn:b9pb853) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ebca8530fb79fc455377afa67f72494309a5f95/pkg/obsservice/obslib/produce/producer.go#L68)):*
> ```Go
> 	}
> 	if len(producers) == 0 {
> 		return nil, errors.New("must provide at least one EventProducer")
> ```

nit: can the alias and owner be added to error message?
___
*[`pkg/obsservice/obslib/queue/mem_queue.go` line 41 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NgKbS3f64J3HzeK-IDG:-NgKbS3f64J3HzeK-IDH:bja76hs) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b293d6879a792f2179f7301e759ff9d27257273/pkg/obsservice/obslib/queue/mem_queue.go#L41)):*
> ```Go
> 	mu     struct {
> 		syncutil.Mutex
> 		queue   *list.List
> ```

Can this be the generic list and avoid the type conversion?
___
*[`pkg/obsservice/obslib/queue/mem_queue.go` line 70 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NgKarw3BWShxmp0a62d:-NgKarw3BWShxmp0a62e:b-rktj2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b293d6879a792f2179f7301e759ff9d27257273/pkg/obsservice/obslib/queue/mem_queue.go#L70)):*
> ```Go
> func (q *MemoryQueue[T]) Enqueue(e T) error {
> 	q.mu.Lock()
> 	defer q.mu.Unlock()
> ```

should the lock be taken after the size function to reduce time it's held?

**Review by abarganier — COMMENTED** [review]
Thanks @j82w - RFAL again

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-:-NgLF1WL3i6_dpaAzCBc:b-9asusv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @j82w, @knz, and @koorosh)

___
*[`pkg/obsservice/obslib/produce/producer.go` line 68 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NgK2XxP0bgzGXzx-2rm:-NgLDIJhEP1XdbfYLiOz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ebca8530fb79fc455377afa67f72494309a5f95/pkg/obsservice/obslib/produce/producer.go#L68)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

nit: can the alias and owner be added to error message?
</blockquote></details>

Done.
___
*[`pkg/obsservice/obslib/queue/mem_queue.go` line 41 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NgKbS3f64J3HzeK-IDG:-NgLDkZOCOPs7SV9mnuT:bfmbs3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b293d6879a792f2179f7301e759ff9d27257273/pkg/obsservice/obslib/queue/mem_queue.go#L41)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

Can this be the generic list and avoid the type conversion?
</blockquote></details>

I don't think such a thing exists in the stdlib unfortunately. It wraps elements with the `Element` type under the hood and I don't see an option to use generics, nor do I want to put in the effort this early on to implement the underlying list from scratch.

If a time comes where this is a problem, I figure we can optimize then. 
___
*[`pkg/obsservice/obslib/queue/mem_queue.go` line 70 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-NgKarw3BWShxmp0a62d:-NgLE6lE4IHQx5FnW9vi:bz5py0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b293d6879a792f2179f7301e759ff9d27257273/pkg/obsservice/obslib/queue/mem_queue.go#L70)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

should the lock be taken after the size function to reduce time it's held?
</blockquote></details>

Good idea! Done.

**Review by j82w — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107191#-:-NgM-BsX-iLO247yslXo:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian, @knz, and @koorosh)

**Comment by abarganier** [comment]
TFTR!

bors r=j82w

**Pull request merged**